### ISSUE TITLE
Game list building improvements

### ISSUE THREAD

**Issue by nullpainter** [mika76/mamesaver#6]
On my laptop (i7, 2.9Ghz, 32GB RAM), building the game list consumes 3Gb of RAM and takes almost 7 minutes. Whereas this is possibly unavoidable for the first run, it becomes tedious having to rebuild the entire game list every time a new game is added.

A couple of suggestions:

1. If possible, some manner of progress dialog would be useful. Not sure if we can guess the expected size of the output XML file and provide a semi-accurate progress dialog. 
2. A separate 'Rescan games' button. This would require the full XML game list being stored alongside the existing `gamelist.xml`. Instead of running MAME to generate the full list again, it would just rely on the XML file already downloaded. This way, the full list rebuild would only be required on the very first time or if MAME was upgraded.

I haven't performed any analysis as to why it's taking so long or consuming so much RAM. Generating the list from the command line and piping to a file takes 30 seconds. I know that XPath isn't that performant on large XML files (and perhaps my slightly lazy `//machine` XPath may not be helping).

If you don't have the time or inclination to do this work, I may be able to submit a pull request.

**Comment by nullpainter** [comment]
Changing the `//machine` XPath to `mame/machine` reduces the elapsed time to a bit over three minutes. I've only done a single run on both so this isn't hugely scientific, however a more specific XPath query should theoretically be more performant. 

If we ignore the separate button suggestion for the time-being, I'm happy to do a proper performance test and identify the bottleneck.

**Comment by mika76** [comment]
Hey thanks for the interest @nullpainter!

Yeah the build list functionality is a pain.  Options I thought of...

1. Looking at http://docs.mamedev.org/commandline/commandline-all.html I could probably pull the list using -listfull which comes back within a second if piped, but there is no year included only the game name.

2. One idea I had was to actually get the rom folders by using the `-showconfig` param, get the files list there, and then get the `-listxml` info for only the roms in the folders. This would probably have a bug effect for most people since I doubt most people have 1000's of roms.

3. One other issue is the fact that if I pipe the `-listxml` for all games it's like 430MB! I didn't really want such a huge cached file permanently saved. Of course there is the option to just pull out the info needed (after the list has been pulled - so still quite a while) and then cache only that info (rom name, game name and year really). If doing this it would probably be good to store the hash of the mame.exe and compare so we know if it was changed/updated. This would save time on subsequent tries but not on the first one.

Either opt 1 or 2 sound the best to me. 1 if showing the year is dropped (seems silly to do all this to just show year), or 2 if year is important.

> If you don't have the time or inclination to do this work, I may be able to submit a pull request.

I'm very happy to receive contributions 😊 Please just try keep them small and focused one thing at a time - last one I got changed everything including formatting and had the binaries included!

**Comment by nullpainter** [comment]
Good point about the size of the full XML file; I had forgotten that! Given that piping directly to a file from mame.exe takes under a minute, I may start by rewriting the parsing using `XMLReader` instead of XPath to see how much this alleviates the problem.

Even though it pains me, I promise to keep the project using .NET Framework 2.0 ;-)

**Comment by mika76** [comment]
> Even though it pains me, I promise to keep the project using .NET Framework 2.0 ;-)

Hey I'm quite happy to update to a newer version of .net - update away 😄

**Comment by nullpainter** [comment]
Good news. By reading directly from the `StandardOutput` stream instead of first reading the contents into a string, and using a combination of `XmlReader` traversal with LINQ to XML to simplify the parsing, I've reduced the `-listxml` parsing from 4.5 minutes and 3Gb of RAM (!) down to 26 seconds and 33Mb of RAM. This is about the amount of time that the `-listxml` command itself takes.

Unfortunately the `-verifyroms` step still takes just under three minutes - again because the MAME command takes that long. 😢 

So, this is a good first step that I think should be done regardless of any future smarts.

I've upgraded to .NET 4.6 in order to use the LINQ to XML package and am tempted to shift all of this to a separate class. It would also be nice to be able maintain previous game selections after scanning, so there could potentially be quite a bit of related game preparation code. I've never been a huge fan of #region, however I'm happy to follow your preferred style.

I'll tidy this up and hope to create a PR later this weekend.

**Comment by nullpainter** [comment]
I lied. Added a small degree of parallelisation. Down to two seconds for 15 ROMs.

**Comment by mika76** [comment]
> Definitely finished with this one now

🤣
